### PR TITLE
GSREN3D-487: Solar panel placement performance

### PR DESCRIPTION
**Pull request by ismailsunni**
### JIRA issue

https://jira.camptocamp.com/browse/GSREN3D-487

### Description

- Add unit test for big roof with obstacle solar panel placement
- Investigate the real issue
- Make an improvement

#### Test scenario
- Address: 10 Rue Jean Macé
- Roof: no 10 (not too big not too small)
- With and without obstacles

#### Result

Without obstacles
```
MapComponent.vue:161 roofsStore.saveCleanMatrix(): 162.39999997615814
MapComponent.vue:162 substractSelectedSquaresFromGrid(roofsStore.gridMatrix!): 14
MapComponent.vue:165 solarPanelPlacement(roofsStore.gridMatrix!): 63.10000002384186
MapComponent.vue:166 mapStore.activate3d(): 636.5
MapComponent.vue:167 solarPanelGridToSolarPanelModel: 441.7999999821186
MapComponent.vue:168 Full setupSolarPanel: 1317.8999999761581
MapComponent.vue:213 setupSolarPanel: 1333.9000000059605
MapComponent.vue:214 displaySolarPanelLayer: 4.4000000059604645
MapComponent.vue:215 setupSolarPanel + displaySolarPanelLayer: 1329.5
```
With obstacles
```
MapComponent.vue:161 roofsStore.saveCleanMatrix(): 114.59999999403954
MapComponent.vue:162 substractSelectedSquaresFromGrid(roofsStore.gridMatrix!): 16050.90000000596
MapComponent.vue:165 solarPanelPlacement(roofsStore.gridMatrix!): 73.40000000596046
MapComponent.vue:166 mapStore.activate3d(): 862.2999999821186
MapComponent.vue:167 solarPanelGridToSolarPanelModel: 361.09999999403954
MapComponent.vue:168 Full setupSolarPanel: 17462.5
MapComponent.vue:213 setupSolarPanel: 17469.79999998212
MapComponent.vue:214 displaySolarPanelLayer: 3.399999976158142
MapComponent.vue:215 setupSolarPanel + displaySolarPanelLayer: 17466.40000000596
```

| File                 | Function                                                 | with obstacles | Without Obstacle |
| -------------------- | -------------------------------------------------------- | -------------- | ---------------- |
| MapComponent.vue 161 | roofsStore.saveCleanMatrix()                             | 162.39         | 114.59           |
| MapComponent.vue 162 | **substractSelectedSquaresFromGrid(roofsStore.gridMatrix!)** | 14             | 16050.90         |
| MapComponent.vue 165 | solarPanelPlacement(roofsStore.gridMatrix!)              | 63.10          | 73.40            |
| MapComponent.vue 166 | mapStore.activate3d()                                    | 636.5          | 862.29           |
| MapComponent.vue 167 | solarPanelGridToSolarPanelModel                          | 441.79         | 361.09           |
| MapComponent.vue 168 | Full setupSolarPanel                                     | 1317.89        | 17462.5          |
| MapComponent.vue 213 | setupSolarPanel                                          | 1333.90        | 17469.79         |
| MapComponent.vue 214 | displaySolarPanelLayer                                   | 4.40           | 3.39             |
| MapComponent.vue 215 | setupSolarPanel + displaySolarPanelLayer                 | 1329.5         | 17466.40         |

#### Finding
Based on the log, the most increased time spent is substractSelectedSquaresFromGrid(roofsStore.gridMatrix!), more than 1000x slower.

The issue is not only on the big roof, but also small roof.

We can continue this on the next sprint.

### Screenshots

(only if the final render is different)

### CODE REVIEW

**Comment by ismailsunni** [comment]
Ok, I manage to make it much faster. 

The improvement only needs **~12 seconds** for this kind of obstacle (same address, but use the biggest roof).

image

It took **~1 minute 20 seconds** with less number of obstacles before the improvement

image

**Comment by ismailsunni** [comment]
@MaYohann @toni-garcia-camptocamp I think it's ready to review. Please ignore the `console.log` and `performance.now`, I will remove it before merging this PR. I keep it just in case you want to test it locally (to see the number)

File: src/services/roofInteractionHelper.ts [hunk 1]
```diff
@@ -270,17 +270,28 @@ export function substractSquareFromRoofPanUnion(roofPans: Feature<Geometry>) {
...
+    }
     for (y = 0; y < squareGrid[x].length; y++) {
       for (const selectedSquare of selectedSquares) {
-        const center = selectedSquare.getProperty('center') as Point
-        if (booleanEqual(center, squareGrid[x][y].squareCenter as Point)) {
-          squareGrid[x][y].usable = false
-          break
+        // Only check if the square is usable because we want to set the
+        // usable to false if it equals
+        if (squareGrid[x][y].usable) {
```
[the remark this answers is not in the document]
Comment (ismailsunni) [reply src/services/roofInteractionHelper.ts:286]: ah yes, indeed. Thanks!

**Review by toni-garcia-camptocamp — APPROVED** [review]
LGTM

**Pull request merged**